paddle/framework/channel.h

```diff
@@ -13,75 +13,52 @@ See the License for the specific language governing permissions and
 limitations under the License. */
 
 #pragma once
-#include <condition_variable>
-#include <mutex>
-#include <queue>
+
+#include <stddef.h>  // for size_t
 
 namespace paddle {
 namespace framework {
 
+// Channel is the abstract class of buffered and un-buffered channels.
 template <typename T>
 class Channel {
  public:
-  explicit Channel(std::size_t capacity) : capacity_(capacity) {}
-
-  void Send(T* channel_element) {
-    std::unique_lock<std::mutex> lock(mu_);
-
-    if (IsBounded()) {
-      full_cond_var_.wait(lock, [this]() {
-        bool capacity_valid = capacity_ > 0 ? !IsCapacityFull() : true;
-        return capacity_valid;
-      });
-    }
-    channel_.push_back(std::move(*channel_element));
-
-    lock.unlock();
-    empty_cond_var_.notify_one();
-  }
-
-  T* Receive() {
-    std::unique_lock<std::mutex> lock(mu_);
-    empty_cond_var_.wait(lock, [this]() { return !channel_.empty(); });
+  virtual void Send(T*) = 0;
+  virtual T* Receive() = 0;
+  virtual size_t Cap() = 0;
 
-    T* channel_element = std::move(channel_.front());
-    channel_.pop_front();
+  // Don't delete channels; instead, call Channel::Close.
+ protected:
+  virtual ~Channel() {}
+};
 
-    NotifyAllSenders(&lock);
-    return channel_element;
-  }
 
-  size_t Size() {
-    std::unique_lock<std::mutex> lock(mu_);
-    return channel_.size();
-  }
+// Forward declaration of channel implementations.
+namespace details {
+template <typename T> class Buffered;
+template <typename T> class UnBuffered;
+}  // namespace details
 
-  void Clear() {
-    std::unique_lock<std::mutex> lock(mu_);
-    channel_.clear();
 
-    NotifyAllSenders(&lock);
+template <typename T>
+Channel<T>* MakeChannel(size_t buffer_size) {
```

Comment:
I think we should move these functions to a separate file like `channel_utils`. The current design forces us to put includes at the end of the header file to avoid cyclic dependencies, which is not a good practice. 
https://stackoverflow.com/questions/19514786/include-at-the-end-of-a-header-file-is-it-safe-good-practise

Reply:
Given that the link you posted says:

> However, if the includes are considered to be implementation details that would only be a distraction, they could be placed at the bottom.

it seems neutral to have header files included at the bottom of a header file.

Comparing the cyclic inclusion and the forcing of users to write two includes

```c++
#include "paddle/framework/channel.h"
#include "paddle/framework/channel_utils.h"
``

I'd choose the former one.